mini-projects/00-jq-rs/src/main.rs

```diff
@@ -1,3 +1,112 @@
-fn main() {
-    println!("Hello World!");
+use anyhow::{Context, Result};
+use clap::Parser;
+use serde_json::Value;
+use std::{
+    env,
+    io::{stdout, BufReader},
+    num::ParseIntError,
+    path::PathBuf,
+};
+use thiserror::Error;
+
+mod filters;
+use filters::pipe;
+
+mod functions;
+mod output;
+use output::print_result;
+
+#[derive(Error, Debug)]
+enum MyErrors {
+    #[error("The indent value must be in the range of 0-7 inclusive. Current value: {0}")]
+    InvalidIndent(usize),
+
+    #[error("You can have either colored or monochrome output")]
+    ColoredAndMonochromeError,
+
+    #[error("You can have either compact or indented output")]
+    CompactAndIndentedError,
+
+    #[error("Failed to read the provided JSON file: {0}")]
+    JSONError(#[from] serde_json::Error),
+
+    #[error("You need an integer")]
+    ParseError(#[from] ParseIntError),
+}
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[clap(short, long, default_value_t = true)]
+    color_output: bool,
+
+    #[clap(short, long)]
+    monochrome_output: bool,
+
+    #[clap(short, long)]
+    sort_keys: bool,
+
+    #[clap(short, long, default_value_t = 2)]
+    indent: usize,
+
+    #[clap(short = 'o', long)]
+    compact_output: bool,
+
+    #[clap(required = true)]
+    needle: String,
+
+    #[clap(required = true)]
+    file: PathBuf,
+}
+
+pub fn file_path(path: PathBuf) -> Result<Value> {
+    let file =
+        std::fs::File::open(&path).with_context(|| format!("Failed to open file: {:?}", path))?;
+    let reader = BufReader::new(file);
+    let json_value: Value = serde_json::from_reader(reader)?;
+
+    Ok(json_value)
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+
+    let jq_colors = env::var("JQ_COLORS").unwrap_or_else(|_| {
+        if !args.color_output || args.monochrome_output {
+            "0;0:0;0:0;0:0;0:0;0:0;0:0;0:0;0".to_string()
```

Comment:
Would have been nice to have this as a less opaque datatype - one example is having the actual colors and formats as n explicit hash map to what they map to

As it stands, it would be really annoying for someone to change of the types, for example if I wanted to change null to be printed red, I would have to look at the source code for how this gets parsed to know which attribute to change, rather than a well-named entry in a hash map